DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/README.md

```diff
@@ -0,0 +1,286 @@
+   > ```
+   > constexpr size_t kNumBanks = fpga_tools::Pow2(fpga_tools::CeilLog2(kNumRows));
+   > ```
+
+3. Change the loop induction variables to match with the memory access pattern and result in the correct indices to `array2d` as assigned in change 1.
+
+The `OptimizedKernel` implements these changes, and executes the following memory access pattern:
+
+![Optimized Kernel Memory Access Pattern](./assets/access_pattern_optimized.gif)
+
+Note that in the Kernel Memory Viewer, although the same bank bits ($b_4$, $b_3$, $b_2$) are used, this time, the memory partitioning is exactly what desired. We can deduce a memory layout using the same table:
+
+|              | Address | Bank bits | Bank # |
+|--------------|---------|-----------|--------|
+| array2d[0][0] | 0x0000  | 0b000     | 0      |
+| array2d[0][1] | 0x0004  | 0b001     | 1      |
+| array2d[0][2] | 0x0008  | 0b010     | 2      |
+| array2d[0][3] | 0x000C  | 0b011     | 3      |
+| array2d[0][4] | 0x0010  | 0b100     | 4      |
+| array2d[0][5] | 0x0014  | 0b010     | 5      |
+| array2d[0][6] | 0x0018  | 0b011     | 6      |
+| array2d[0][7] | 0x001C  | 0b100     | 7      |
+| array2d[1][0] | 0x0020  | 0b000     | 0      |
+| array2d[1][1] | 0x0024  | 0b001     | 1      |
+| ...           |         |           |        |
+
+Because that we are only accessing the first 5 columns of `array2d`, that provides sufficient knowledge for the compiler to optimize out the unused memory banks - bank 5 to 7. As a result, we have the following memory layout:
+
+![Optimized Kernel Memory Layout](./assets/memory_system_optimized.png)
+
+Review the memory access pattern above, we can easily find out that within each loop iteration, each of the memory banks is access precisely twice: 1 load and 1 store. Note that though the `idx` loop also performs loads on each bank, since the values to be loaded are already accessed in the `col` loop, these loads are shared. This time, the memory accesses do not surpass the physical limit of the memory block hardware, and thus, we have eliminated the memory bottleneck. In the meantime, observe that the Kernel Memory Viewer shows pipelined never-stall LSUs for our optimized memory system, and the compiler successfully optimized out the unused memory banks.
+
+![Never-stall Node](./assets/memory_report_optimized.png)
+
+Subsequently, since we have made the LSU non-stallable, we expect no stalls in the loop nor in the downstream task - that the output pipe `OutStream_OptKernel` is not stalled. To verify this improvement, refer to the simulation waveform and observe the `valid` and `data` signal of the output pipe `ID_OutStream_OptKernel_pipe_channel_write`. The gaps no longer appear in the `valid` signal, and the pipe writes out data every clock cycle.
+
+![Optimized System Waveform](./assets/waveform_optimized.png)
```

Comment:
the font in these questa screenshots is hard to read. try and capture a narrower field (that is, squeeze more samples in) like this:

![image](https://github.com/user-attachments/assets/302f3dd6-e383-4c2e-bc12-3a97f7827a6f)